homeassistant/components/weather/__init__.py

```diff
@@ -114,52 +190,125 @@ class WeatherEntity(Entity):
     _attr_precision: float
-    _attr_pressure: float | None = None
+    _attr_native_pressure: float | None = None
     _attr_pressure_unit: str | None = None
+    _attr_native_pressure_unit: str | None = None
     _attr_state: None = None
-    _attr_temperature_unit: str
-    _attr_temperature: float | None
-    _attr_visibility: float | None = None
+    _attr_temperature_unit: str | None = None
+    _attr_native_temperature_unit: str
+    _attr_native_temperature: float
+    _attr_native_visibility: float | None = None
     _attr_visibility_unit: str | None = None
+    _attr_native_visibility_unit: str | None = None
+    _attr_native_precipitation: float | None = None
     _attr_precipitation_unit: str | None = None
+    _attr_native_precipitation_unit: str | None = None
     _attr_wind_bearing: float | str | None = None
-    _attr_wind_speed: float | None = None
+    _attr_native_wind_speed: float | None = None
     _attr_wind_speed_unit: str | None = None
+    _attr_native_wind_speed_unit: str | None = None
+
+    _weather_option_temperature_uom: str | None = None
+    _weather_option_pressure_uom: str | None = None
+    _weather_option_visibility_uom: str | None = None
+    _weather_option_precipitation_uom: str | None = None
+    _weather_option_wind_speed_uom: str | None = None
+
+    _deprecated_weather_entity_report = False
+    _deprecated_weather_entity_reported = False
+
+    def __init_subclass__(cls, **kwargs: Any) -> None:
+        """Post initialisation processing."""
+        super().__init_subclass__(**kwargs)
+        if "temperature_unit" in cls.__dict__:
+            cls._deprecated_weather_entity_report = True
+
+    async def async_internal_added_to_hass(self) -> None:
+        """Call when the sensor entity is added to hass."""
+        await super().async_internal_added_to_hass()
+        if not self.registry_entry:
+            return
+        self.async_registry_entry_updated()
 
     @property
-    def temperature(self) -> float | None:
+    def native_temperature(self) -> float:
         """Return the platform temperature in native units (i.e. not converted)."""
-        return self._attr_temperature
+        return self._attr_native_temperature
 
     @property
-    def temperature_unit(self) -> str:
+    def native_temperature_unit(self) -> str | None:
         """Return the native unit of measurement for temperature."""
-        return self._attr_temperature_unit
+        if hasattr(self, "_attr_native_temperature_unit"):
+            return self._attr_native_temperature_unit
+        return None
 
     @property
-    def pressure(self) -> float | None:
+    def temperature_unit(self) -> str:
+        """Return the unit of measurement for temperature."""
+        if (
+            self._deprecated_weather_entity_report is True
+            and self._deprecated_weather_entity_reported is False
+        ):
+            self._report_deprecated_weather_entity()
+
+        if self.native_temperature_unit is None:
```

Comment:
Use walrus operator here, and in the other  unit methods
```suggestion
        if (native_temperature_unit := self.native_temperature_unit) is None:
```